examples/operator.rs

```diff
@@ -0,0 +1,100 @@
+extern crate timely;
+extern crate timely_communication;
+
+use std::collections::HashMap;
+
+use timely_communication::Configuration;
+use timely::dataflow::operators::{ToStream, Input, Inspect, Operator, FrontierNotificator};
+use timely::dataflow::channels::pact::Pipeline;
+use timely::progress::timestamp::RootTimestamp;
+use timely::dataflow::Scope;
+
+fn main() {
+    timely::execute(Configuration::Thread, |root| {
+        root.scoped(|scope| {
+            (0u64..10).to_stream(scope)
+                .unary_frontier(Pipeline, "example", |mut builder| {
+                    let mut cap = Some(builder.get_cap(RootTimestamp::new(12)));
+                    move |(input, _frontier), output| {
+                        cap = None;
+                        while let Some((time, data)) = input.next() {
+                            output.session(&time).give_content(data);
+                        }
+                    }
+                }).inspect(|x| println!("{:?}", x));
+        });
+    }).unwrap();
+
+    timely::execute(Configuration::Thread, |root| {
+        root.scoped(|scope| {
+            (0u64..10).to_stream(scope)
+                .unary(Pipeline, "example", |mut builder| {
```

Comment:
Another option (other than requesting capabilities) is just to be given a capability for the default (least, I hope) time. You can then `delayed` it to whatever times you need, or just drop it if you don't want it. This is done at initialization time, so it shouldn't be terribly expensive.

You could imagine writing something like

```rust
stream.unary(exchange, "example", |cap| {}, |input, output| {
    // logic ..
});
```

where if you'd like an initial notification, you do something with `cap` in that closure. 

This has the annoying property that your notificator would be borrowed by both closures, and so you would need to `Rc<RefCell<_>>` it up. Perhaps the constructor approach you are suggesting is the tasteful way to get around that? In which case, we could simplify it by removing the "request" part and just always giving you a `Capability::new(Default::default())` or whatever the least time is. You would write semi-gross things like:

```rust
stream.unary(exchange, "example", |cap| { |input, output| {
    // logic ..
}});
```

which maybe isn't horrible? I haven't tried, but it would be cool to know if all the ownership works out ok, or if this ends up a pain.

Having written a few weird operators ("weird": not much use for traditional notificator), I think what I'd love to see is: 

1. Frontier for each input as part of `InputHandle`. This seems to make so much sense. For operators that don't request frontier information, we just keep it empty (or make it an `Option<&[Timestamp]>`? Seems like a pain)
2. `Notificator` demoted to user code; you should only get one if you mint it up and use it.
3. Perhaps "capability list" associated with outputs, dual to the frontiers for inputs? I'm usually just writing against one output and holding my capabilities separately; perhaps I should keep doing that.

How do these strike you?